File: apps/api/src/modules/public/public.controller.ts
import { Controller, Get, Param, Query, NotFoundException } from '@nestjs/common';
import { PublicService } from './public.service';

@Controller('public')
export class PublicController {
  constructor(private readonly publicService: PublicService) {}

  @Get('jobs')
  async listJobs(
    @Query('q') q?: string,
    @Query('province') province?: string,
    @Query('tradeId') tradeId?: string,
    @Query('site') site?: string,
    @Query('page') page?: string,
    @Query('limit') limit?: string,
    @Query('lat') lat?: string,
    @Query('lng') lng?: string,
    @Query('radiusKm') radiusKm?: string,
    @Query('statusFilter') statusFilter?: string,
    @Query('minWage') minWage?: string,
    @Query('maxWage') maxWage?: string,
  ) {
    const validStatus = ['CLOSING_SOON', 'CLOSED'] as const;
    const parsedStatus = validStatus.includes(statusFilter as typeof validStatus[number])
      ? (statusFilter as typeof validStatus[number])
      : undefined;
    return this.publicService.listJobs({
      q: q?.trim() || undefined,
      province,
      tradeId: tradeId ? Number(tradeId) : undefined,
      site,
      page: page ? Number(page) : 1,
      limit: limit ? Math.min(50, Number(limit)) : undefined,
      lat: lat ? Number(lat) : undefined,
      lng: lng ? Number(lng) : undefined,
      radiusKm: radiusKm ? Math.min(200, Number(radiusKm)) : undefined,
      statusFilter: parsedStatus,
      minWage: minWage ? Number(minWage) : undefined,
      maxWage: maxWage ? Number(maxWage) : undefined,
    });
  }

  @Get('jobs/:slug')
  async getJobBySlug(@Param('slug') slug: string) {
    const job = await this.publicService.getJobBySlug(slug);
    if (!job) throw new NotFoundException(`Job not found: ${slug}`);
    return job;
  }

  @Get('sites/:slug')
  async getSiteBySlug(@Param('slug') slug: string) {
    const site = await this.publicService.getSiteById(slug);
    if (!site) throw new NotFoundException(`Site not found: ${slug}`);
    return site;
  }

  @Get('provinces')
  async getProvinces() {
    return this.publicService.getProvinces();
  }

  @Get('trades')
  async getTrades() {
    return this.publicService.getTrades();
  }
}

File: apps/api-kotlin/src/main/kotlin/vn/gada/api/public_/PublicController.kt
package vn.gada.api.public_

import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import vn.gada.api.common.exception.NotFoundException

@RestController
@RequestMapping("/public")
class PublicController(private val publicService: PublicService) {

    /** GET /public/jobs */
    @GetMapping("/jobs")
    fun listJobs(
        @RequestParam(required = false) province: String?,
        @RequestParam(required = false) tradeId: String?,
        @RequestParam(required = false) site: String?,
        @RequestParam(required = false) page: String?,
        @RequestParam(required = false) lat: String?,
        @RequestParam(required = false) lng: String?,
        @RequestParam(required = false) radiusKm: String?,
        @RequestParam(required = false) statusFilter: String?,
        @RequestParam(required = false) minWage: String?,
        @RequestParam(required = false) maxWage: String?,
        @RequestParam(required = false) minExp: String?
    ): ResponseEntity<Map<String, Any?>> {
        val validStatuses = setOf("CLOSING_SOON", "CLOSED")
        val parsedStatus = if (statusFilter != null && validStatuses.contains(statusFilter)) statusFilter else null
        val validExpFilters = setOf("none", "lt1", "1to2", "2to3", "gte3")
        val parsedExp = if (minExp != null && validExpFilters.contains(minExp)) minExp else null

        val params = mapOf<String, Any?>(
            "province" to province,
            "tradeId" to tradeId?.toIntOrNull(),
            "site" to site,
            "page" to (page?.toIntOrNull() ?: 1),
            "lat" to lat?.toDoubleOrNull(),
            "lng" to lng?.toDoubleOrNull(),
            "radiusKm" to radiusKm?.let { minOf(200.0, it.toDoubleOrNull() ?: 50.0) },
            "statusFilter" to parsedStatus,
            "minWage" to minWage?.toLongOrNull(),
            "maxWage" to maxWage?.toLongOrNull(),
            "minExp" to parsedExp
        )

        return ok(publicService.listJobs(params))
    }

    /** GET /public/wage-stats */
    @GetMapping("/wage-stats")
    fun getWageStats(): ResponseEntity<Map<String, Any?>> {
        return ok(publicService.getWageStats())
    }

    /** GET /public/jobs/:slug */
    @GetMapping("/jobs/{slug}")
    fun getJobBySlug(@PathVariable slug: String): ResponseEntity<Map<String, Any?>> {
        val job = publicService.getJobBySlug(slug)
            ?: throw NotFoundException("Job not found: $slug")
        return ok(job)
    }

    /** GET /public/sites/:slug */
    @GetMapping("/sites/{slug}")
    fun getSiteBySlug(@PathVariable slug: String): ResponseEntity<Map<String, Any?>> {
        val site = publicService.getSiteById(slug)
            ?: throw NotFoundException("Site not found: $slug")
        return ok(site)
    }

    /** GET /public/provinces */
    @GetMapping("/provinces")
    fun getProvinces(): ResponseEntity<Map<String, Any?>> {
        return ok(publicService.getProvinces())
    }

    /** GET /public/trades */
    @GetMapping("/trades")
    fun getTrades(): ResponseEntity<Map<String, Any?>> {
        return ok(publicService.getTrades())
    }

    private fun ok(data: Any?): ResponseEntity<Map<String, Any?>> =
        ResponseEntity.ok(mapOf("statusCode" to 200, "data" to data))
}
